src/bucket/BucketInputIterator.cpp

```diff
@@ -87,7 +87,8 @@ BucketInputIterator::getMetadata() const
     return mMetadata;
 }
 
-BucketInputIterator::BucketInputIterator(std::shared_ptr<Bucket const> bucket)
+BucketInputIterator::BucketInputIterator(std::shared_ptr<Bucket const> bucket,
+                                         std::string const& inFilename)
```

Comment:
I've changed the internals of `Bucket` to have an optional filename/hash for each sort order. This means that a bucket can now have just the old sort order, just the new sort order, both, or neither (representing the default, empty `Bucket` object). However, I still have `Bucket::getFilename(BucketSortOrder) -> fs::path` and `Bucket::getHash(BucketSortOrder) -> Hash`. Currently the codebase treats empty string and `Hash{}` as null values, so I have the functions return those values if the underlying class member is `std::nullopt`. Is it worth refactoring the codebase to support actual optionals instead of these null values? It's a fairly large refactor so I wanted to double check before doing it. Other than having to copy values instead of returning a reference, I don't see any other drawback for optional values. However I wanted to make sure that there wasn't some advantage or edge case I was missing making an empty `Hash` better than an `std::optional<Hash>`.

Reply:
I think it's a worthwhile refactor to try in the future -- along with ever-more migrations from std::strings to std::filesystem::path -- but I wouldn't gate this change on it, you've got enough going on in here already, and IIRC we special-case the empty / zero hash in a number of places in the codebase, not just buckets.

Reply:
Actually as I read it and fiddle around with the code I'm increasingly nervous that we're conflating "empty hash" with important state transitions here. It might be worth refactoring to make the optionals explicit in more places. I'm going to give it a try to see how terrible it is, might add that last bit if it seems ok.